crates/bevy_ecs/src/component.rs

```diff
@@ -261,7 +279,70 @@ impl ComponentInfo {
 
     /// Create a new [`ComponentInfo`].
     pub(crate) fn new(id: ComponentId, descriptor: ComponentDescriptor) -> Self {
-        ComponentInfo { id, descriptor }
+        ComponentInfo {
+            id,
+            descriptor,
+            hooks: ComponentHooks::default(),
+        }
+    }
+
+    /// Register a [`ComponentHook`] that will be run when this component is added to an entity
+    ///
+    /// Will panic if the component already has an `on_add` hook
```

Comment:
These docs could be clearer on what the difference between `on_add` and `on_insert` are. When would one get called and not the other? Is there an overlap to when they're called or are they mutually exclusive? Some specific examples of when each is and is not called would be incredibly helpful.

Reply:
Yes, this could use additional documentation, for reference `on_insert` will run after any call to `.insert` but `on_add` will only run if the entity didn't already have the component.

Reply:
`on_insert` also gets called when the entity is spawned right?

Reply:
Yes, `on_add` will never trigger without being followed by `on_insert`.

Reply:
What's the motivation for including `on_insert`? Could we treat re-insertions the same as just overwriting a mutable reference?

Reply:
`on_insert` exists in lieu of being able to actually trigger hooks for mutations. In cases where you want to update some index structure with new values it's possible to either manually enforce only mutating via insertion, or enforce it at the type level by hiding component types.